zcash_client_sqlite/src/wallet.rs

```diff
@@ -536,6 +541,98 @@ pub(crate) fn block_height_extrema(
     })
 }
 
+fn parse_block_metadata(
+    row: (BlockHeight, Vec<u8>, Option<u32>, Vec<u8>),
+) -> Option<Result<BlockMetadata, SqliteClientError>> {
+    let (block_height, hash_data, sapling_tree_size_opt, sapling_tree) = row;
+    let sapling_tree_size = sapling_tree_size_opt.map(Ok).or_else(|| {
+        if sapling_tree == BLOCK_SAPLING_FRONTIER_ABSENT {
+            None
+        } else {
+            // parse the legacy commitment tree data
+            read_commitment_tree::<
+                zcash_primitives::sapling::Node,
+                _,
+                { zcash_primitives::sapling::NOTE_COMMITMENT_TREE_DEPTH },
+            >(Cursor::new(sapling_tree))
+            .map(|tree| Some(tree.size().try_into().unwrap()))
+            .map_err(SqliteClientError::from)
+            .transpose()
+        }
+    })?;
```

Comment:
Hm. Maybe I should just make it an error at this level if we attempt to parse block metadata and have no possible source for the note commitment tree size. I didn't really like collapsing the two cases together, and after the migration really the only time that block metadata should be non-present is when the wallet is empty, since even for linear scanning we'll be persisting the note commitment tree sizes.